matrix_content_scanner/scanner/scanner.py

```diff
@@ -114,9 +124,77 @@ async def scan_file(
             FileDirtyError if the result of the scan said that the file is dirty, or if
                 the media path is malformed.
         """
-        # Compute the cache key for the media.
+        # Compute the key to use when caching, both in the current scans cache and in the
+        # results cache.
         cache_key = self._get_cache_key_for_file(media_path, metadata, thumbnail_params)
+        if cache_key not in self._current_scans:
+            # Create a future in the context of the current event loop.
+            loop = asyncio.get_event_loop()
+            f = loop.create_future()
+            # Register the future in the current scans cache so that subsequent queries
+            # can use it.
+            self._current_scans[cache_key] = f
+            # Try to download and scan the file.
+            try:
+                res = await self._scan_file(
+                    cache_key, media_path, metadata, thumbnail_params
+                )
+                # Set the future's result, and mark it as done.
+                f.set_result(res)
+                # Return the result.
+                return res
+            except Exception as e:
+                # If there's an exception, catch it, pass it on to the future, and raise
+                # it.
+                f.set_exception(e)
+                # We retrieve the exception from the future, because if we don't and no
+                # other request is awaiting on the future, asyncio complains about "Future
+                # exception was never retrieved".
+                f.exception()
+                raise
+            finally:
+                # Remove the future from the cache.
+                del self._current_scans[cache_key]
+
+        # We need to turn the future into a Deferred before awaiting on it, to make
+        # Twisted and asyncio happy.
+        d: Deferred[MediaDescription] = Deferred.fromFuture(
```

Comment:
We need this extra annotation due to https://github.com/twisted/twisted/issues/11753